polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -924,6 +974,36 @@ async fn process_incoming_peer_message<Context>(
 	}
 }
 
+// Holds off an AssetHub collation, if it's not from an invulnerable AssetHub collator. Returns
+// `true` if the collation was held off and `false` otherwise.
+fn hold_off_asset_hub_collation_if_needed(
+	state: &mut State,
+	origin: PeerId,
+	relay_parent: Hash,
+	prospective_candidate: Option<(CandidateHash, Hash)>,
+) -> bool {
+	// If we don't know the peer we should reject the advertisement but to avoid verbosity and
+	// copy-pasted logic we'll just return `false` and let the caller handle it.
+	let maybe_para_id = state.peer_data.get(&origin).and_then(|pd| pd.collating_para());
+	if maybe_para_id == Some(ASSET_HUB_PARA_ID) && !state.ah_invulnerables.contains(&origin) {
+		state.ah_held_off_collations.push_back((
```

Comment:
What If a collator advertises the same collation multiple times ? it seems it there are no checks. In `handle_advertisement` we actually check and return `InsertAdvertisementError::Duplicate` .